### ISSUE TITLE
The instrumentation hook is only called after visiting a route

### ISSUE THREAD

**Issue by daveyjones** [vercel/next.js#59999]
### Link to the code that reproduces this issue

https://github.com/daveyjones/nextjs-instrumentation-bug

### To Reproduce

```shell
git clone git@github.com:daveyjones/nextjs-instrumentation-bug.git
cd nextjs-instrumentation-bug
npm install
npm run dev # The register function IS called
npm run build && npm start # The register function IS NOT called until you visit http://localhost:3000
```

### Current vs. Expected behavior

The `register` function should be called automatically after running `npm run build && npm start`. You should not have to visit a route for the function to run.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.1.0: Mon Oct  9 21:28:12 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T8103
Binaries:
  Node: 20.9.0
  npm: 10.2.4
  Yarn: 1.22.19
  pnpm: N/A
Relevant Packages:
  next: 14.0.5-canary.29
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router

### Additional context

This behavior is also reported in the comments of #49897 ([reference](https://github.com/vercel/next.js/issues/49897#issuecomment-1822377705)). However, since #49897 is focused on `standalone` mode (which is not a factor here), I felt it was best to create a separate issue. It may be the case that resolving this issue will also resolve #49897.

**Comment by ships** [comment]
I arrived at this issue after knocking my head on a related problem, but which turns out to be different -- leaving comment here for other seekers.

My issue was that instrumentation hooks were not running at all in a docker context (without standalone) even though they were running on `next start` locally.

`next build` puts the instrumentation hook into the built directory, but still relies on `next.config.js` to specify the experimental flag for those hooks to run. In my context I had copied the build directory into my docker image but left `next.config.js` behind.

After keeping `next.config.js` around, I was able to get the hooks running, but they still run only once requests are received. However, I don't have the issue of it happening repeatedly -- `register` are only invoked on the first API route handled.

**Comment by anuraags** [comment]
This is actually a pretty major bug that took me a while to debug. I had to create a copy of my next.config.js into `next.prod.config.js` that is loaded into the docker context just to get my instrumentation working.

I have no issue with running the register function for every request, but I think the running of the instrumentation should be baked into the `next build` step.
